File: Advanced Sorting Algorithm/ShellSort.java
import java.util.Random;

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public LinkedList() {
        this.head = null;
    }

    // Metode untuk menyisipkan elemen di akhir linked list
    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // Metode untuk mencetak isi linked list
    public void printList() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }

    // Metode untuk menghitung panjang linked list
    public int getLength() {
        int count = 0;
        Node current = head;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    // Metode untuk mendapatkan node ke-n dalam linked list
    public Node getNth(int n) {
        Node current = head;
        for (int i = 0; i < n && current != null; i++) {
            current = current.next;
        }
        return current;
    }

    public void addNode(int i) {
    }

    public void display() {
    }
}

//Proses algoritma shell sorting //
public class ShellSort {                                     
    public static void shellSort(LinkedList list) {                  
        int n = list.getLength();
        int gap = n / 2;

        while (gap > 0) {
            for (int i = gap; i < n; i++) {
                int temp = list.getNth(i).data;                     
                int j = i;

                while (j >= gap && list.getNth(j - gap).data > temp) {
                    list.getNth(j).data = list.getNth(j - gap).data;
                    j -= gap;
                }

                list.getNth(j).data = temp;
            }
            gap /= 2;
        }
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        Random random = new Random();

        // Menambahkan 10 angka acak antara 1 dan 20 ke dalam linked list
        for (int i = 0; i < 100; i++) {
            int randomNumber = random.nextInt(1000) + 1;
            list.insertAtEnd(randomNumber);
        }

        System.out.println("Linked List sebelum diurutkan:");
        list.printList();

        shellSort(list);

        System.out.println("\nLinked List setelah diurutkan:");
        list.printList();
    }
}

File: Advanced Sorting Algorithm/Shell Sort.py
import random

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insertAtEnd(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def printList(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next
        print()

    def getLength(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count

    def getNth(self, n):
        current = self.head
        for i in range(n):
            if current is None:
                return None
            current = current.next
        return current

def shellSort(linked_list):
    n = linked_list.getLength()
    gap = n // 2

    while gap > 0:
        for i in range(gap, n):
            temp = linked_list.getNth(i).data
            j = i

            while j >= gap and linked_list.getNth(j - gap).data > temp:
                linked_list.getNth(j).data = linked_list.getNth(j - gap).data
                j -= gap

            linked_list.getNth(j).data = temp
        gap //= 2

if __name__ == "__main__":
    linked_list = LinkedList()

    # Menambahkan 100 angka acak antara 1 dan 100 ke dalam linked list tanpa angka yang sama
    random_numbers = random.sample(range(1, 101), 100)
    for number in random_numbers:
        linked_list.insertAtEnd(number)

    print("Linked List sebelum diurutkan:")
    linked_list.printList()

    shellSort(linked_list)

    print("\nLinked List setelah diurutkan:")
    linked_list.printList()
